### PR TITLE
Move ASAN to clang12 and Ubuntu-22.04 (Jammy)

### PR DESCRIPTION
**Pull request by malfet**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106355
* #106354

- Modify `install_conda` to remove libstdc++ from libstdcxx-ng to use one from OS
- Modify `install_torchvision` to workaround weird glibc bug, where malloc interposers (such as ASAN) are causing a hang in internationalization library, see https://sourceware.org/bugzilla/show_bug.cgi?id=27653 and https://gcc.gnu.org/bugzilla/show_bug.cgi?id=90589
- Modify `torch.utils.cpp_extension` to recognize Ubuntu's clang as supported compiler

Extracted from https://github.com/pytorch/pytorch/pull/105260

### CODE REVIEW

**Comment by malfet** [comment]
I expect lintrunner will complaint that slow has not been migrated (even though they should be in sync..)

File: .ci/docker/common/install_conda.sh [hunk 1]
```diff
@@ -109,7 +109,8 @@ if [ -n "$ANACONDA_PYTHON_VERSION" ]; then
   # gcc-9 for ubuntu-18.04 from http://ppa.launchpad.net/ubuntu-toolchain-r/test/ubuntu
   # Pulls llibstdc++6 13.1.0-8ubuntu1~18.04 which is too new for conda
   # So remove libstdc++6.so.3.29 installed by https://anaconda.org/anaconda/libstdcxx-ng/files?version=11.2.0
-  if grep 18.04.6 /etc/issue >/dev/null; then
+  # Same is true for gcc-12 from Ubuntu-22.04
+  if grep -e [12][82].04.[62] /etc/issue >/dev/null; then
```
Comment (huydhn) [inline .ci/docker/common/install_conda.sh:113]: lol, this is wicked

**Review by huydhn — APPROVED** [review]
There is also an ASAN slow job that needs update https://github.com/pytorch/pytorch/blob/main/.github/workflows/slow.yml#L110-L131

File: .ci/docker/common/install_conda.sh (hunk 1 above)
Comment (malfet) [reply .ci/docker/common/install_conda.sh:113]: It matches 18.04.6 and 22.04.2, and everything in between :)

**Comment by malfet** [comment]
For posterity: `test_cpp_wrapper` dynamo jobs fail with ambiguous constructor error:
```
/tmp/torchinductor_jenkins/py39_cpu/inline_extension_cwnqbbq5lr6hsaktauqhm5hulaxgvvwxphzkz3docrqablnmbd4v/main.cpp:41:50: error: call to constructor of 'c10::ArrayRef<c10::IValue>' is ambiguous
        RECORD_FUNCTION("inductor_wrapper_call", c10::ArrayRef<c10::IValue>({{}}));
                                                 ^                          ~~~~
/var/lib/jenkins/workspace/torch/include/ATen/record_function.h:580:38: note: expanded from macro 'RECORD_FUNCTION'
      at::RecordScope::FUNCTION, fn, inputs, ##__VA_ARGS__)
                                     ^~~~~~
/var/lib/jenkins/workspace/torch/include/ATen/record_function.h:561:20: note: expanded from macro 'RECORD_FUNCTION_WITH_SCOPE'
        guard, fn, inputs, ##__VA_ARGS__);                 \
                   ^~~~~~
/var/lib/jenkins/workspace/torch/include/c10/util/ArrayRef.h:40:7: note: candidate constructor (the implicit move constructor)
class ArrayRef final {
      ^
/var/lib/jenkins/workspace/torch/include/c10/util/ArrayRef.h:40:7: note: candidate constructor (the implicit copy constructor)
/var/lib/jenkins/workspace/torch/include/c10/util/ArrayRef.h:71:13: note: candidate constructor
  constexpr ArrayRef(const T& OneElt) : Data(&OneElt), Length(1) {}
            ^
/var/lib/jenkins/workspace/torch/include/c10/util/ArrayRef.h:126:28: note: candidate constructor
  /* implicit */ constexpr ArrayRef(const std::initializer_list<T>& Vec)
                           ^
1 error generated.
```

**Pull request closed without merging**